sw/airborne/modules/dragspeed/dragspeed.c

```diff
@@ -0,0 +1,228 @@
+			/ dragspeed.coeff.y;
+	// Simple low-pass filter
+	dragspeed.vel.x += (1 - dragspeed.filter) * (vx - dragspeed.vel.x);
+	dragspeed.vel.y += (1 - dragspeed.filter) * (vy - dragspeed.vel.y);
+	// Send as ABI VELOCITY_ESTIMATE message
+	// Note: set VEL_DRAGSPEED_ID to ABI_DISABLE to disable
+#if VEL_DRAGSPEED_ID
+	if (!dragspeed.do_calibrate_coeff && !dragspeed.do_calibrate_zero) {
+		AbiSendMsgVELOCITY_ESTIMATE(VEL_DRAGSPEED_ID, stamp,
+				dragspeed.vel.x, dragspeed.vel.y, 0, DRAGSPEED_R);
+	}
+#endif
+	// Perform calibration if required
+	calibrate_coeff(accel);
+	calibrate_zero(accel);
+}
+
+/**
+ * Calibrate drag coefficient by comparing accelerometer measurements to INS
+ * velocities.
+ * Should be performed with VEL_DRAGSPEED_ID set to ABI_DISABLE, but for safety
+ * the ABI messages are also disabled automatically when calibration is active.
+ *
+ * This routine assumes that the accelerometers have been zeroed beforehand.
+ */
+static void calibrate_coeff(struct Int32Vect3 *accel) {
+	// Reset when new calibration is started
+	static int do_calibrate_prev = 0;
+	static struct FloatVect2 coeff;
+	static int num_samples_x = 0;
+	static int num_samples_y = 0;
+	if (dragspeed.do_calibrate_coeff && !do_calibrate_prev) {
+		coeff.x = 0;
+		coeff.y = 0;
+		num_samples_x = 0;
+		num_samples_y = 0;
+	}
+	do_calibrate_prev = dragspeed.do_calibrate_coeff;
+	// Return when calibration is not active
+	if (!dragspeed.do_calibrate_coeff) {
+		return;
+	}
+
+	// Calculate INS velocity in body frame
+	struct FloatEulers *att = stateGetNedToBodyEulers_f();
+	struct EnuCoor_f *vel_ins = stateGetSpeedEnu_f();
+	struct FloatVect2 vel_ins_body = {
+			cos(att->psi) * vel_ins->y + sin(att->psi) * vel_ins->x,
+			-sin(att->psi) * vel_ins->y + cos(att->psi) * vel_ins->x
+	};
+	// Calibrate coefficient when velocity is sufficiently high
+	if (abs(vel_ins_body.x) > 0.5) {
+		float this_coeff = -ACCEL_FLOAT_OF_BFP(accel->x) / vel_ins_body.x;
+		coeff.x = (coeff.x * num_samples_x + this_coeff) / (num_samples_x + 1);
+		num_samples_x++;
+	}
+	if (abs(vel_ins_body.y) > 0.5) {
+		float this_coeff = -ACCEL_FLOAT_OF_BFP(accel->y) / vel_ins_body.y;
+		coeff.y = (coeff.y * num_samples_y + this_coeff) / (num_samples_y + 1);
+		num_samples_y++;
+	}
+	// End calibration when enough samples are averaged
+	if (num_samples_x > 1000 && num_samples_y > 1000 && coeff.x != 0
+			&& coeff.y != 0) {
+		dragspeed.coeff = coeff;
+		dragspeed.do_calibrate_coeff = 0;
+	}
+}
+
+/**
+ * Calibrate zero velocity by measuring the accelerations while the drone
+ * hovers in-place.
+ *
+ * The zero-velocity readings can change between flights, e.g. because of small
+ * differences in battery or outer hull positions.
+ */
+static void calibrate_zero(struct Int32Vect3 *accel) {
+	// Reset when new calibration is started
+	static int do_calibrate_prev = 0;
+	static struct FloatVect2 zero;
+	static int num_samples = 0;
+	if (dragspeed.do_calibrate_zero && !do_calibrate_prev) {
+		zero.x = 0;
```

Comment:
0.f

Small thing but it helps to reinforce that this variable is a float, as opposed to the num_samples.